File: cses/09_geometry/2195_convex-hull/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    coords = data[1:1 + 2 * n]
    xs = list(map(int, coords[0::2]))
    ys = list(map(int, coords[1::2]))
    pts = sorted(zip(xs, ys))  # sắp xếp các điểm theo (x, y)

    if n <= 1:
        out = [str(n)]
        out.extend("%d %d" % (x, y) for x, y in pts)
        sys.stdout.write("\n".join(out) + "\n")
        return

    # Lower hull (duyệt trái -> phải): GIỮ LẠI điểm collinear -> chỉ pop khi
    # cross < 0 (góc quẹo phải thực sự).
    # Tối ưu hằng số: theo dõi độ dài bằng m, unpack đỉnh stack ra biến local.
    lower = []
    m = 0
    lpop = lower.pop
    lapp = lower.append
    for p in pts:
        px = p[0]
        py = p[1]
        while m >= 2:
            ax, ay = lower[m - 2]
            bx, by = lower[m - 1]
            # cross(a, b, p) = (b - a) x (p - a), số nguyên nên chính xác tuyệt đối
            if (bx - ax) * (py - ay) - (by - ay) * (px - ax) < 0:
                lpop()
                m -= 1
            else:
                break
        lapp(p)
        m += 1

    # Upper hull: duyệt ngược (phải -> trái), cùng điều kiện pop cross < 0.
    upper = []
    m = 0
    upop = upper.pop
    uapp = upper.append
    for p in reversed(pts):
        px = p[0]
        py = p[1]
        while m >= 2:
            ax, ay = upper[m - 2]
            bx, by = upper[m - 1]
            if (bx - ax) * (py - ay) - (by - ay) * (px - ax) < 0:
                upop()
                m -= 1
            else:
                break
        uapp(p)
        m += 1

    # Bỏ điểm cuối của mỗi chain (trùng với đỉnh đầu của chain kia) rồi ghép lại.
    hull = lower[:-1] + upper[:-1]

    out = [str(len(hull))]
    out.extend("%d %d" % (x, y) for x, y in hull)
    sys.stdout.write("\n".join(out) + "\n")


main()

File: cses/09_geometry/2195_convex-hull/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

struct Point {
    long long x;
    long long y;

    // Sắp xếp các điểm theo thứ tự (x, y) tăng dần.
    bool operator<(const Point& other) const {
        if (x != other.x) {
            return x < other.x;
        }
        return y < other.y;
    }
};

// Cross product nguyên (exact integer): dương -> quẹo trái, âm -> quẹo phải,
// bằng 0 -> ba điểm thẳng hàng (collinear). Dùng long long nên không tràn số.
long long cross(const Point& a, const Point& b, const Point& c) {
    return (b.x - a.x) * (c.y - a.y) -
           (b.y - a.y) * (c.x - a.x);
}

// Dựng một nửa hull theo Andrew's monotone chain. Chỉ pop khi cross < 0 (góc
// quẹo phải THỰC SỰ) để GIỮ LẠI các điểm collinear nằm trên cạnh biên.
std::vector<Point> build_half(const std::vector<Point>& points) {
    std::vector<Point> half;
    half.reserve(points.size());
    for (const Point& point : points) {
        while (half.size() >= 2 &&
               cross(half[half.size() - 2], half.back(), point) < 0) {
            half.pop_back();
        }
        half.push_back(point);
    }
    return half;
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::cin >> n;
    std::vector<Point> points(static_cast<std::size_t>(n));
    for (Point& point : points) {
        std::cin >> point.x >> point.y;
    }
    std::sort(points.begin(), points.end());

    // CSES bảo đảm diện tích hull dương. Xử lý thêm trường hợp suy biến (tất cả
    // điểm thẳng hàng) để tránh hai monotone chain in trùng các điểm collinear
    // nếu điều kiện đó về sau bị nới lỏng.
    bool all_collinear = true;
    for (std::size_t i = 2; i < points.size(); ++i) {
        if (cross(points[0], points[1], points[i]) != 0) {
            all_collinear = false;
            break;
        }
    }
    if (all_collinear) {
        std::cout << points.size() << '\n';
        for (const Point& point : points) {
            std::cout << point.x << ' ' << point.y << '\n';
        }
        return 0;
    }

    // Lower hull duyệt trái -> phải; upper hull duyệt phải -> trái (đảo mảng).
    std::vector<Point> lower = build_half(points);
    std::reverse(points.begin(), points.end());
    std::vector<Point> upper = build_half(points);

    // Bỏ điểm cuối của mỗi chain (trùng với đỉnh đầu của chain kia) rồi ghép lại.
    std::vector<Point> hull;
    hull.reserve(lower.size() + upper.size() - 2);
    hull.insert(hull.end(), lower.begin(), lower.end() - 1);
    hull.insert(hull.end(), upper.begin(), upper.end() - 1);

    std::cout << hull.size() << '\n';
    for (const Point& point : hull) {
        std::cout << point.x << ' ' << point.y << '\n';
    }
    return 0;
}
